### PR TITLE
Add forward AD gradcheck

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58094 Forward AD formulas batch 3
* #57863 Forward AD formulas batch 2
* #57768 Forward AD formulas batch 1
* #57767 Codegen inplace forward AD formula from out of place one if needed
* #57701 Add forward AD test for op info
* **#57633 Add forward AD gradcheck**

Differential Revision: [D28387765](https://our.internmc.facebook.com/intern/diff/D28387765)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
         jacobian_cols[d_idx] = _compute_numerical_jvps_wrt_specific_input(jvp_fn, eps, x.is_complex())
     return _combine_jacobian_cols(jacobian_cols, outputs, input, input.numel())
 
+def _get_analytical_jacobian_forward_ad(fn, inputs, outputs, *, all_u=None) -> Tuple[Tuple[torch.Tensor, ...], ...]:
+    """Computes the analytical Jacobian using forward mode AD of `fn(inputs)` with respect
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:289]: Maybe reword as: Computes the analytical Jacobian of `fn(inputs)` using forward-mode AD

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -806,41 +886,88 @@ def wrapped_fn(*inputs):
...
     if not outputs:
         return _check_no_differentiable_outputs(func, tupled_inputs, _as_tuple(func_out), eps)
 
     numerical = _transpose(_get_numerical_jacobian(func, tupled_inputs, outputs, eps=eps))
 
-    for i, o in enumerate(outputs):
-        analytical = _check_analytical_jacobian_attributes(tupled_inputs, o, nondet_tol, check_grad_dtypes)
-        inp_tensors = _iter_tensors(tupled_inputs, True)
+    if use_forward_ad:
+        complex_indices = [i for i, inp in enumerate(tupled_inputs) if inp.is_complex()]
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:952]: Do we need to compute complex_indices again?

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
...
+    with fwAD.dual_level():
+        fw_grads = []
+        new_inputs = []
+        for i, inp in enumerate(inputs):
+            if is_tensor_like(inp) and inp.requires_grad:
+                if inp.layout == torch._mkldnn:  # type: ignore[attr-defined]
+                    raise ValueError("MKLDNN inputs are not support for forward AD gradcheck.")
+
+                inp = fwAD.make_dual(inp, torch.zeros_like(inp))
+                # If inp is a differentiable view, the dual might no be the tangent given to
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:326]: no -> not

File: torch/autograd/gradcheck.py [hunk 4]
```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
...
+
+    if all_u:
+        dummy_reduced_inputs = tuple(torch.tensor(1., requires_grad=i.requires_grad) for i in inputs)
+        jacobians = tuple(_allocate_jacobians_with_inputs(dummy_reduced_inputs, o.numel()) for o in outputs)
+    else:
+        jacobians = tuple(_allocate_jacobians_with_inputs(inputs, o.numel()) for o in outputs)
+
+    with fwAD.dual_level():
+        fw_grads = []
+        new_inputs = []
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:319]: Maybe a more descriptive name would be new_inputs -> dual_inputs?

File: torch/autograd/gradcheck.py [hunk 5]
```diff
@@ -806,41 +886,88 @@ def wrapped_fn(*inputs):
...
+            return _as_tuple(fn(*new_inputs))
+        return wrapped_fn
+    return apply_to_c_inps(fn, lambda x: x + 0 * 1j), apply_to_c_inps(fn, lambda x: x * 1j)
+
 
 def _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps, rtol,
-                         atol, check_grad_dtypes, nondet_tol, complex_indices, any_outputs_complex):
-    if any_outputs_complex:
-        real_fn, imag_fn = _real_and_imag(func, outputs)
+                         atol, check_grad_dtypes, check_forward_ad, nondet_tol):
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:901]: Feels like there's some logic repeated here between the two blocks, but not entirely sure there's an easy way to fix it.

File: torch/autograd/gradcheck.py [hunk 6]
```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
...
+        all_u (optional): if provided, the Jacobian will be right multiplied with this vector
+
+    Returns:
+        A list of M N-tuples of tensors
+    """
+    # To avoid early import issues
+    fwAD = torch.autograd.forward_ad
+
+    if any(i.is_complex() for i in inputs):
+        raise ValueError("Expected inputs to be non-complex for _get_analytical_jacobian_forward_ad.")
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:309]: We can probably add some tests here to check if all of these errors can be reached
ditto for the one below for MKLDNN inputs

File: torch/autograd/gradcheck.py [hunk 7]
```diff
@@ -806,41 +886,88 @@ def wrapped_fn(*inputs):
...
 
+def _real_and_imag_input(fn):
+    # returns new functions that take real inputs instead of complex inputs and compute fn(x + 0 * 1j)
+    # and f(x * 1j).
+    def apply_to_c_inps(fn, fn_to_apply):
+        def wrapped_fn(*inputs):
+            new_inputs = (fn_to_apply(inp) if inp.is_complex() else inp for inp in inputs)
+            return _as_tuple(fn(*new_inputs))
+        return wrapped_fn
+    return apply_to_c_inps(fn, lambda x: x + 0 * 1j), apply_to_c_inps(fn, lambda x: x * 1j)
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:897]: This is an interesting way of converting number to complex `x + 0 * 1j`, did you just do this for symmetry (because the other lambda had `x * 1j`)?

**Review by soulitzer — COMMENTED** [review]
Looks pretty good overall! Added some minor comments.

**Comment by soulitzer** [comment]
In terms of further testing, what if you had a wrapper function that unpacked the returned dual tensor from the original function, perturbs the tangent by some eps, and then does make_dual again before returning?

File: torch/autograd/gradcheck.py [hunk 8]
```diff
@@ -806,41 +894,88 @@ def wrapped_fn(*inputs):
...
+
 
 def _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps, rtol,
-                         atol, check_grad_dtypes, nondet_tol, complex_indices, any_outputs_complex):
-    if any_outputs_complex:
-        real_fn, imag_fn = _real_and_imag(func, outputs)
+                         atol, check_grad_dtypes, check_forward_ad, nondet_tol):
+    complex_indices = [i for i, o in enumerate(outputs) if o.is_complex()]
+    if complex_indices:
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/autograd/gradcheck.py]: Yes I saw that. Fixed :)

File: torch/autograd/gradcheck.py (hunk 2 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:952]: These indices are for the inputs.
Should I use different names to make this obvious?

File: torch/autograd/gradcheck.py (hunk 5 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:901]: I don't think so, because we now wrap the input and not the output and we need to pass slightly different flags.
I feel like abstracting that would just create more boilerplate code.

File: torch/autograd/gradcheck.py (hunk 7 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:897]: Not, that's because that way, I don't need to worry about real -> float type promotion for float32 vs float64 :D And so I don't need a manual mapping of type promotions :D

File: test/test_autograd.py [hunk 9]
```diff
@@ -4297,6 +4297,29 @@ def fn2(x):
...
+        y = torch.rand(2, dtype=torch.double, requires_grad=True)
+
+        gradcheck(fn, (x, y), check_forward_ad=True, fast_mode=False)
+        gradcheck(fn, (x, y), check_forward_ad=True, fast_mode=True)
+
+        # Test for one input and one output being complex
+        x = torch.rand(2, dtype=torch.cdouble, requires_grad=True)
+
+        gradcheck(fn, (x, y), check_forward_ad=True, fast_mode=False)
+        gradcheck(fn, (x, y), check_forward_ad=True, fast_mode=True)
```
Comment (zou3519) [inline test/test_autograd.py:4315]: Both the outputs here are complex, right?

File: test/test_autograd.py [hunk 10]
```diff
@@ -4297,6 +4297,29 @@ def fn2(x):
         check(fast_mode=True)
         check(fast_mode=False)
 
+    def test_gradcheck_forward_ad(self):
```
Comment (zou3519) [inline test/test_autograd.py:4300]: Probably want to test a failure case to catch things like accidental early returns

File: torch/autograd/gradcheck.py (hunk 6 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:309]: The output validation is shared, so no need to test anything. But adding a test for this input check indeed.

File: torch/autograd/gradcheck.py [hunk 11]
```diff
@@ -285,6 +285,80 @@ def get_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, ep
...
+    using the passed ones (many torch.nn tests do this).
+
+    Args:
+        fn: the function to compute the jacobian for
+        inputs: inputs to `fn`
+        outputs: provide precomputed outputs to avoid one extra invocation of fn
+        all_u (optional): if provided, the Jacobian will be right multiplied with this vector
+
+    Returns:
+        A list of M N-tuples of tensors
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:303]: nit: type signature says Tuple[Tuple[Tensor, ...], ...], should this be "a tuple of M N-tuples"?

File: torch/autograd/gradcheck.py [hunk 12]
```diff
@@ -806,41 +886,88 @@ def wrapped_fn(*inputs):
...
-    for i, o in enumerate(outputs):
-        analytical = _check_analytical_jacobian_attributes(tupled_inputs, o, nondet_tol, check_grad_dtypes)
-        inp_tensors = _iter_tensors(tupled_inputs, True)
+    if use_forward_ad:
+        complex_indices = [i for i, inp in enumerate(tupled_inputs) if inp.is_complex()]
+        analytical_forward = _get_analytical_jacobian_forward_ad(func, tupled_inputs, outputs)
+
+        for i, (n_per_out, a_per_out) in enumerate(zip(numerical, analytical_forward)):
+            for j, (n, a) in enumerate(zip(n_per_out, a_per_out)):
+                if a.numel() != 0 or n.numel() != 0:
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:957]: Shouldn't torch.allclose be able to compare tensors that have 0 numel? Is this check redundant? (a comment would be nice if there is a reason)

File: test/test_autograd.py (hunk 9 above)
Comment (albanD) [reply test/test_autograd.py:4315]: Good catch! The second return of `f` should be `y`. updated.

File: torch/autograd/gradcheck.py (hunk 12 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:957]: I just re-used existing code here. Not sure.

File: test/test_autograd.py (hunk 10 above)
Comment (albanD) [reply test/test_autograd.py:4300]: Added with Jeffrey suggestion above to unpack/repack with noisy gradients.

File: torch/autograd/gradcheck.py (hunk 12 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:957]: Oh I think it can be removed. I originally added it only for debugging purposes, when I wanted to compute `(a - n).abs().max()`, and reduction ops require non-zero numel.

File: torch/autograd/gradcheck.py (hunk 2 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:952]: Oh maybe something like complex_inp_indices/complex_out_indices would be helpful

File: torch/autograd/gradcheck.py (hunk 12 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:957]: Updated.

**Comment by albanD** [comment]
Sorry for the important update but implementing the OpInfo showed some limitations in the implementation that was here and that it wasn't working for most complex workloads.
This updated version works unchanged when OpInfos are added and won't get any other major modification.
If some edge cases I don't know about are not supported here, they will be added in a follow up PR.

File: test/test_autograd.py [hunk 13]
```diff
@@ -4297,6 +4314,58 @@ def fn2(x):
...
+        gradcheck(fn, (x, y), check_forward_ad=True, fast_mode=True)
+        with self.assertRaisesRegex(RuntimeError, err_msg):
+            gradcheck(bad_fn, (x, y), check_forward_ad=True, fast_mode=False)
+        with self.assertRaisesRegex(RuntimeError, err_msg):
+            gradcheck(bad_fn, (x, y), check_forward_ad=True, fast_mode=True)
+
+        # Test for all inputs and outputs being complex
+        y = torch.rand(2, dtype=torch.cdouble, requires_grad=True)
+
+        gradcheck(fn, (x, y), check_forward_ad=True, fast_mode=False)
```
Comment (soulitzer) [inline test/test_autograd.py:4293]: maybe wrap all the checks in a function and call it with fast_mode=True/False

**Comment by albanD** [comment]
This is ready for final review

File: torch/autograd/gradcheck.py [hunk 14]
```diff
@@ -223,7 +227,7 @@ def _combine_jacobian_cols(jacobians_cols: Dict[int, List[torch.Tensor]], output
                            numel) -> Tuple[torch.Tensor, ...]:
     # jacobian_cols maps column_idx -> output_idx -> single column of jacobian Tensor
     # we return a list that maps output_idx -> full jacobian Tensor
-    jacobians = _allocate_jacobians_with_outputs(outputs, numel, input.dtype, input.device)
+    jacobians = _allocate_jacobians_with_outputs(outputs, numel, dtype=input.dtype if input.dtype.is_complex else None)
```
Comment (soulitzer) [inline torch/autograd/gradcheck.py:230]: Previously this function allocated tensors to have the dtype and device of the inputs (but after the change, the outputs), so not sure if you meant to make this change.

**Review by soulitzer — COMMENTED** [review]
Added a couple more comments. Might be good to make a quick PR testing that CI passes with slow_mode=False if you haven't already.

File: torch/autograd/gradcheck.py (hunk 14 above)
Comment (albanD) [reply torch/autograd/gradcheck.py:230]: This function was always called with real valued outputs. And so the dtype of the input was always the thing to use.
Now when the input is real but output is complex, we want to use the output dtype.
So this doesn't change the behavior of existing code, but allows for complex output.

**Comment by soulitzer** [comment]
Yeah, that was what I was thinking. 
Also @gchanan says that we should allow people to test slow_mode in ci-all

Alternatively though, we can have a new branch prefix specifically for new gradcheck
See https://github.com/pytorch/pytorch/pull/57880

**Comment by albanD** [comment]
@soulitzer Running the slow gradcheck version in https://github.com/pytorch/pytorch/pull/57952

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387765).

**Review by soulitzer — APPROVED** [review]
LGTM! Thanks for the addition!

**Comment by mruberry** [comment]
Unlanding this stack as it appears to have hit a logical merge conflict and is breaking multiple builds on the HUD. Sample tests broken:

test_forward_mode_AD_narrow_cpu_complex128
test_forward_mode_AD_narrow_cpu_float64
test_forward_mode_AD_acos_cuda_complex128
test_forward_mode_AD_acosh_cuda_complex128
test_forward_mode_AD_narrow_cuda_complex128
test_forward_mode_AD_narrow_cuda_float64
